File: intellifile_app/lib/sync/vector_clock.dart
// lib/sync/vector_clock.dart
//
// Vector clock for conflict resolution.
// Tracks which device last edited a file and detects concurrent edits.
// Uses sqflite for persistence on mobile.

import 'dart:convert';
import 'package:sqflite/sqflite.dart';
import 'package:path/path.dart' as p;
import 'package:path_provider/path_provider.dart';

const String kDeviceId = 'mobile';

// ─── VectorClock data class ─────────────────────────────────────────────────

class VectorClock {
  final String deviceId;
  Map<String, double> clock;

  VectorClock({required this.deviceId, Map<String, double>? clock})
      : clock = clock ?? {};

  /// Tick this device's entry — call after every local file edit.
  void tick() {
    clock[deviceId] =
        DateTime.now().millisecondsSinceEpoch / 1000.0;
  }

  /// Merge a remote clock into this one — take max per device.
  void merge(Map<String, dynamic> other) {
    for (final entry in other.entries) {
      final remoteTs = (entry.value as num).toDouble();
      clock[entry.key] = (clock[entry.key] ?? 0) > remoteTs
          ? clock[entry.key]!
          : remoteTs;
    }
  }

  /// Compare against a remote clock.
  ///
  /// Returns one of:
  /// - `local_wins`  — our clock is strictly ahead
  /// - `remote_wins` — remote clock is strictly ahead
  /// - `identical`   — both clocks are equal
  /// - `conflict`    — both sides have edits since last sync
  String compare(Map<String, dynamic> other) {
    final allDevices = <String>{...clock.keys, ...other.keys};

    bool localNewer = false;
    bool remoteNewer = false;

    for (final device in allDevices) {
      final localTs = clock[device] ?? 0.0;
      final remoteTs = (other[device] as num?)?.toDouble() ?? 0.0;

      if (localTs > remoteTs) localNewer = true;
      if (remoteTs > localTs) remoteNewer = true;
    }

    if (localNewer && !remoteNewer) return 'local_wins';
    if (remoteNewer && !localNewer) return 'remote_wins';
    if (!localNewer && !remoteNewer) return 'identical';
    return 'conflict';
  }

  Map<String, dynamic> toJson() => Map<String, dynamic>.from(clock);

  factory VectorClock.fromJson(String deviceId, Map<String, dynamic> json) {
    return VectorClock(
      deviceId: deviceId,
      clock: json.map((k, v) => MapEntry(k, (v as num).toDouble())),
    );
  }
}

// ─── SQLite persistence ─────────────────────────────────────────────────────

class VectorClockStore {
  static Database? _db;

  static Future<Database> _getDb() async {
    if (_db != null) return _db!;
    final dir = await getApplicationDocumentsDirectory();
    final dbPath = p.join(dir.path, 'intellifil_mobile.db');
    _db = await openDatabase(
      dbPath,
      version: 1,
      onCreate: (db, version) async {
        await db.execute('''
          CREATE TABLE IF NOT EXISTS vector_clocks (
            filepath    TEXT PRIMARY KEY,
            clock_json  TEXT NOT NULL,
            updated_at  REAL NOT NULL
          )
        ''');
      },
    );
    return _db!;
  }

  /// Load the vector clock for a file. Returns a fresh clock if not found.
  static Future<VectorClock> loadClock(String filepath) async {
    final db = await _getDb();
    final rows = await db.query(
      'vector_clocks',
      where: 'filepath = ?',
      whereArgs: [filepath],
    );

    if (rows.isNotEmpty) {
      final clockJson =
          jsonDecode(rows.first['clock_json'] as String) as Map<String, dynamic>;
      return VectorClock.fromJson(kDeviceId, clockJson);
    }
    return VectorClock(deviceId: kDeviceId);
  }

  /// Save a vector clock for a file.
  static Future<void> saveClock(String filepath, VectorClock vc) async {
    final db = await _getDb();
    await db.insert(
      'vector_clocks',
      {
        'filepath': filepath,
        'clock_json': jsonEncode(vc.clock),
        'updated_at': DateTime.now().millisecondsSinceEpoch / 1000.0,
      },
      conflictAlgorithm: ConflictAlgorithm.replace,
    );
  }

  /// Load all vector clocks — sent during handshake.
  static Future<Map<String, dynamic>> loadAllClocks() async {
    final db = await _getDb();
    final rows = await db.query('vector_clocks');
    final result = <String, dynamic>{};
    for (final row in rows) {
      result[row['filepath'] as String] =
          jsonDecode(row['clock_json'] as String);
    }
    return result;
  }

  /// Close the database (call on app dispose).
  static Future<void> close() async {
    await _db?.close();
    _db = null;
  }
}

File: sync/vector_clock.py
# pc/sync/vector_clock.py

import json
import logging
import sqlite3
import time
from dataclasses import dataclass, field


DEVICE_ID = "pc"  # change to "mobile" on mobile side

log = logging.getLogger("intellifil.vector_clock")


def _to_int(value) -> int:
    try:
        return int(value)
    except (TypeError, ValueError):
        return 0


def _get_conn(db_path: str) -> sqlite3.Connection:
    conn = sqlite3.connect(db_path, timeout=10)
    try:
        conn.execute("PRAGMA journal_mode=WAL")
        conn.execute("PRAGMA synchronous=NORMAL")
    except sqlite3.Error as exc:
        log.warning("SQLite PRAGMA failed: %s", exc)
    return conn


@dataclass
class VectorClock:
    device_id: str
    clock: dict = field(default_factory=dict)

    def tick(self):
        """Call this every time the local device edits a file."""
        self.clock[self.device_id] = _to_int(self.clock.get(self.device_id, 0)) + 1

    def merge(self, other: dict):
        """Merge a remote clock into this one — take max per device."""
        for device, ts in other.items():
            self.clock[device] = max(_to_int(self.clock.get(device, 0)), _to_int(ts))

    def compare(self, other: dict) -> str:
        """
        Compare this clock against a remote clock.
        Returns one of: 'local_wins' | 'remote_wins' | 'identical' | 'conflict'
        """
        local_newer = any(
            _to_int(self.clock.get(d, 0)) > _to_int(other.get(d, 0))
            for d in set(self.clock) | set(other)
        )
        remote_newer = any(
            _to_int(other.get(d, 0)) > _to_int(self.clock.get(d, 0))
            for d in set(self.clock) | set(other)
        )

        if local_newer and not remote_newer:
            return "local_wins"
        elif remote_newer and not local_newer:
            return "remote_wins"
        elif not local_newer and not remote_newer:
            return "identical"
        else:
            return "conflict"  # both edited since last sync

    def to_json(self) -> str:
        return json.dumps(self.clock)

    @staticmethod
    def from_json(device_id: str, data: str) -> "VectorClock":
        vc = VectorClock(device_id)
        vc.clock = json.loads(data)
        return vc


# ─── Persistence (SQLite) ──────────────────────────────────────────────────────

def init_clock_db(db_path: str):
    """Create the vector_clocks table if it doesn't exist."""
    conn = _get_conn(db_path)
    conn.execute("""
        CREATE TABLE IF NOT EXISTS vector_clocks (
            filepath    TEXT PRIMARY KEY,
            clock_json  TEXT NOT NULL,
            updated_at  REAL NOT NULL
        )
    """)
    conn.commit()
    conn.close()


def save_clock(db_path: str, filepath: str, vc: VectorClock):
    conn = _get_conn(db_path)
    conn.execute("""
        INSERT INTO vector_clocks (filepath, clock_json, updated_at)
        VALUES (?, ?, ?)
        ON CONFLICT(filepath) DO UPDATE SET
            clock_json = excluded.clock_json,
            updated_at = excluded.updated_at
    """, (filepath, vc.to_json(), time.time()))
    conn.commit()
    conn.close()


def load_clock(db_path: str, filepath: str) -> VectorClock:
    """Load clock for a file. Returns a fresh clock if not found."""
    conn = _get_conn(db_path)
    row = conn.execute(
        "SELECT clock_json FROM vector_clocks WHERE filepath=?",
        (filepath,)
    ).fetchone()
    conn.close()

    if row:
        return VectorClock.from_json(DEVICE_ID, row[0])
    return VectorClock(DEVICE_ID)  # new file, fresh clock


def load_all_clocks(db_path: str) -> dict:
    """Load all clocks — sent to remote device during sync handshake."""
    conn = _get_conn(db_path)
    rows = conn.execute(
        "SELECT filepath, clock_json FROM vector_clocks"
    ).fetchall()
    conn.close()
    return {row[0]: json.loads(row[1]) for row in rows}